daemon/api_themes.go

```diff
@@ -0,0 +1,247 @@
+	IconThemes  map[string]themeStatus `json:"icon-themes"`
+	SoundThemes map[string]themeStatus `json:"sound-themes"`
+}
+
+func getInstalledThemes(d *Daemon) (gtkThemes, iconThemes, soundThemes []string, err error) {
+	infos := d.overlord.InterfaceManager().Repository().Info(&interfaces.InfoOptions{
+		Names: []string{"content"},
+		Slots: true,
+	})
+	for _, info := range infos {
+		for _, slot := range info.Slots {
+			var content string
+			// The content interface ensures this attribute exists
+			if err := slot.Attr("content", &content); err != nil {
+				return nil, nil, nil, err
+			}
+			var themes *[]string
+			switch content {
+			case "gtk-3-themes":
+				themes = &gtkThemes
+			case "icon-themes":
+				themes = &iconThemes
+			case "sound-themes":
+				themes = &soundThemes
+			default:
+				continue
+			}
+			var sources []interface{}
+			if err := slot.Attr("source.read", &sources); err != nil {
+				continue
+			}
+			for _, s := range sources {
+				if path, ok := s.(string); ok {
+					*themes = append(*themes, filepath.Base(path))
+				}
+			}
+		}
+	}
+	sort.Strings(gtkThemes)
+	sort.Strings(iconThemes)
+	sort.Strings(soundThemes)
+	return gtkThemes, iconThemes, soundThemes, nil
+}
+
+var badPkgCharRegexp = regexp.MustCompile(`[^a-z0-9]+`)
+
+func themePackageCandidates(prefix, themeName string) []string {
+	themeName = strings.ToLower(themeName)
+	themeName = badPkgCharRegexp.ReplaceAllString(themeName, "-")
+	themeName = strings.Trim(themeName, "-")
+
+	var packages []string
+	for themeName != "" {
+		packages = append(packages, prefix+themeName)
+		pos := strings.LastIndexByte(themeName, '-')
+		if pos < 0 {
+			break
+		}
+		themeName = themeName[:pos]
+	}
+	return packages
+}
+
+func getThemeStatusForPrefix(ctx context.Context, theStore snapstate.StoreService, user *auth.UserState, prefix string, themes, installed []string, toInstall map[string]bool) (status map[string]themeStatus, err error) {
+	status = make(map[string]themeStatus, len(themes))
+
+	for _, theme := range themes {
+		// Skip duplicates
+		if _, ok := status[theme]; ok {
+			continue
+		}
+		if strutil.SortedListContains(installed, theme) {
+			status[theme] = themeInstalled
+			continue
+		}
+		status[theme] = themeUnavailable
+		for _, name := range themePackageCandidates(prefix, theme) {
+			var info *snap.Info
+			if info, err = theStore.SnapInfo(ctx, store.SnapSpec{Name: name}, user); err == store.ErrSnapNotFound {
+				continue
+			} else if err != nil {
+				return nil, err
+			}
+			// Only mark the theme as available if it has
+			// been published to the stable channel.
+			if info.Channel == "stable" {
+				status[theme] = themeAvailable
+				toInstall[name] = true
+			}
+			break
+		}
+	}
+	return status, nil
+}
+
+func getThemeStatus(ctx context.Context, c *Command, user *auth.UserState, gtkThemes, iconThemes, soundThemes []string) (status themeStatusResponse, toInstall []string, err error) {
```

Comment:
s/get//

my preference instead of the two return values would be for this two take some flags let's the caller ask only for available snaps (not installed or unavailable)

Reply:
still agreeing with myself over this